src/Microsoft.Azure.Mobile.Client/Authentication/MobileServicePKCEAuthentication.cs

```
@@ -0,0 +1,88 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+
+using PCLCrypto;
+
+namespace Microsoft.WindowsAzure.MobileServices
+{
+    internal abstract class MobileServicePKCEAuthentication : MobileServiceAuthentication
+    {
+        /// <summary>
+        /// The <see cref="MobileServiceClient"/> used by this authentication session.
+        /// </summary>
+        private readonly MobileServiceClient client;
+
+        protected Uri LoginUri { get; private set; }
+
+        protected Uri CallbackUri { get; private set; }
+        
+        protected string CodeVerifier { get; private set; }
+
+        protected MobileServicePKCEAuthentication(MobileServiceClient client, string provider, string uriScheme, IDictionary<string, string> parameters)
+            : base(client, provider, parameters)
+        {
+            Debug.Assert(client != null, "client should not be null.");
```

+1 for @erychagov - we should fail and throw when uriScheme is invalid. Who's the caller of this constructor? Can you explain what's the behavior of your code here?